docs/src/DeveloperDocumentation/styleguide.md

```diff
@@ -241,3 +217,213 @@ existing block.
+  the concrete instance - exceptions of this rule are possible e.g. for `GF(2)`
+- no nested printing. In particular variable names and `base_ring` must not be displayed.
+  This ensures that `one line` and `:supercompact` stay compact even for complicated things.
+  If you want nested printing use `one line` or `detailed`.
+
+### Implementing show functions
+
+Here is the translation between `:detail`, `one line` and `:supercompact`.
+
+```
+print(io, "text/plain", x)                 # detailed printing
+print(io, x)                               # one line printing
+print(IOContext(:supercompact => true), x) # supercompact printing
+```
+
+For reference, string interpolation `"$(x)"` will also use `print(io, x)`.
+
+#### Mockup
+
+##### Detailed printing with a new line.
+
+```julia
+struct NewRing
+  base_ring
+end
+```
+This implements the detailed printing. Note that the new line is needed for
+this to work. The last print statement must not add a new line.
+
+```julia
+function Base.show(io::IO, ::MIME"text/plain", R::NewRing)
+  println(io, "I am a new ring")
+  println(io, "I print with newlines")
+  print(io, R.base_ring)
+end
+```
+
+This takes care of one line and `:supercompact` printing
+```julia
+function Base.show(io::IO, R::NewRing)
+  if get(io, :supercompact, false)
+    # no nested printing
+    print(io, "supercompact printing of newring ")
+  else
+    # nested printing allowed, preferably supercompact
+    print(io, "one line printing of newring with ")
+    print(IOContext(io, :supercompact => true), "supercompact ", R.base_ring)
+  end
+end
+```
+
+
+##### Detailed printing in a single line.
+
+This version needs to be used in case the detailed
+printing does not contain newlines.
+Then detailed and one line printing agree.
+The if clause takes care of supercompact printing as well.
+```julia
+function Base.show(io::IO, R::NewRing)
+  if get(io, :supercompact, false)
+    # no nested printing
+    print(io, "supercompact printing of newring")
+  else
+    # nested printing allowed, preferably supercompact
+    println(io, "I am a new ring and always print in one line" )
+    print(IOContext(io, :supercompact => true), base_ring(R))
+  end
+end
+```
+
+##### The following is not working as expected and should not be used
+
+```julia
+function Base.show(io::IO, ::MIME"text/plain", R::NewRing)  # do not implement me like this
+  print(io, "I am a new ring with a detailed printing of one line")
+end
+```
+
+Then the following will *not* be used for array/tuple printing.
+It will be used for `print(io, R::NewRing)` though.
+
+```julia
+function Base.show(io::IO, R::NewRing)
+  if get(io, :supercompact, false)
+    print(io, "supercompact printing of newring")
+  else # this is what we call one line
+    print(io, "one line printing of newring with ")
+    print(IOContext(io, :supercompact => true), "supercompact ", R.base_ring)
+  end
+end
+```
+This example illustrates the unexpected behavior.
```

Comment:
I really like that we show the "expected output" here; perhaps we could do the same in the two "good" examples that come before?